packages/3d-web-client-core/src/character/ModelLoader.ts

```diff
@@ -1,68 +1,156 @@
 import { AnimationClip, LoadingManager, Object3D } from "three";
-import { FBXLoader } from "three/examples/jsm/loaders/FBXLoader.js";
-import { GLTF, GLTFLoader } from "three/examples/jsm/loaders/GLTFLoader.js";
+import { GLTF, GLTFLoader as ThreeGLTFLoader } from "three/examples/jsm/loaders/GLTFLoader.js";
+
+class CachedGLTFLoader extends ThreeGLTFLoader {
+  private blobCache: Map<string, string>;
+
+  constructor(manager?: LoadingManager) {
+    super(manager);
+    this.blobCache = new Map();
+  }
+
+  setBlobUrl(originalUrl: string, blobUrl: string) {
+    this.blobCache.set(originalUrl, blobUrl);
+  }
+
+  getBlobUrl(originalUrl: string): string | undefined {
+    return this.blobCache.get(originalUrl);
+  }
+
+  load(
+    url: string,
+    onLoad: (gltf: GLTF) => void,
+    onProgress?: ((event: ProgressEvent<EventTarget>) => void) | undefined,
+    onError?: ((event: ErrorEvent) => void) | undefined,
+  ): void {
+    const blobUrl = this.getBlobUrl(url);
+    if (blobUrl) {
+      console.log(`Loading cached ${url.split("/").pop()}`);
+      super.load(blobUrl, onLoad, onProgress, onError);
+    } else {
+      super.load(url, onLoad, onProgress, onError);
+    }
+  }
+}
+
+class LRUCache<K, V> {
+  private maxSize: number;
+  private cache: Map<K, V>;
+
+  constructor(maxSize: number = 100) {
+    this.maxSize = maxSize;
+    this.cache = new Map();
+  }
+
+  get(key: K): V | undefined {
+    const item = this.cache.get(key);
+    if (item) {
+      this.cache.delete(key);
+      this.cache.set(key, item);
+    }
+    return item;
+  }
+
+  set(key: K, value: V): void {
+    if (this.cache.size >= this.maxSize) {
+      const oldestKey = this.cache.keys().next().value;
+      this.cache.delete(oldestKey);
+    }
+    this.cache.set(key, value);
+  }
+}
+
+interface CachedModel {
+  blob: Blob;
+  originalExtension: string;
+}
 
 export class ModelLoader {
-  private debug: boolean = false;
+  private static instance: ModelLoader | null = null;
+
   private readonly loadingManager: LoadingManager;
-  private readonly fbxLoader: FBXLoader;
-  private readonly gltfLoader: GLTFLoader;
+  private readonly gltfLoader: CachedGLTFLoader;
+  private modelCache: LRUCache<string, CachedModel>;
+  private ongoingLoads: Map<string, Promise<Object3D | AnimationClip | undefined>> = new Map();
 
-  constructor() {
+  constructor(maxCacheSize: number = 100) {
     this.loadingManager = new LoadingManager();
-    this.fbxLoader = new FBXLoader(this.loadingManager);
-    this.gltfLoader = new GLTFLoader(this.loadingManager);
+    this.gltfLoader = new CachedGLTFLoader(this.loadingManager);
+    this.modelCache = new LRUCache(maxCacheSize);
+  }
+
+  /* TODO: decide between below lazy initialization or eager on this file's bottom export */
+  static getInstance(): ModelLoader {
+    if (!ModelLoader.instance) {
+      console.log("Instantiating ModelLoader");
```

Comment:
Remove these `console.log`s

Reply:
Fixed in my local branch about to be pushed on the next PR.